### PR TITLE
Actually setup jit targets when compiling packageimages instead of targeting only one

### LINKED ISSUE #54464: Stdlib pkgimages seems to be compiled with generic target
Making this an issue so I can add it to the milestone:

```julia
julia> cd(joinpath(Sys.BINDIR, "../share/julia/compiled/v1.11/Test"))

julia> Base.parse_image_targets(Base.parse_cache_header("JfdTE_4EB71.ji")[7])
1-element Vector{Base.ImageTarget}:
 generic; flags=0; features_en=(cx16)

julia> VERSION
v"1.11.0-beta1"
```

Ref https://github.com/JuliaLang/julia/pull/54093

### PR DESCRIPTION
**Pull request by gbaraldi**
I haven't assembled a test for this just yet, but this works
Fixes https://github.com/JuliaLang/julia/issues/54464

### CODE REVIEW

File: src/processor_x86.cpp [hunk 1]
```diff
@@ -1084,16 +1082,14 @@ jl_image_t jl_init_processor_pkgimg(void *hdl)
 {
     if (jit_targets.empty())
         jl_error("JIT targets not initialized");
-    if (jit_targets.size() > 1)
```
Comment (vchuravy) [inline src/processor_x86.cpp:1087]: I think we need to test that we still pick the right/one for all.

I have a unsubstantiated fear about loading images with mismatched ABI. Previously this code said to pick one and then ensured that everyone is consistent.

**Comment by vchuravy** [comment]
So checking that my understanding is correct. The issue is that when during sysimg creation we don't yet have a native cache so we are not filling jit targets based on that. Instead we actually parse the `-C` string.

But for pkgimages we have an image loaded, so `ensure_jit_targets` early exists and so the multi versioning infrastructure doesn't see the rest?

Maybe we should split parsing for output and choosing for loading/hit?

**Comment by gbaraldi** [comment]
So the code in `ensure_jit_targets` only seems to actually do anything when building an image, so maybe we should move it around?

**Comment by giordano** [comment]
With the script at https://github.com/JuliaLang/julia/issues/54464#issuecomment-2110544009 I get
```
% JULIA_CPU_TARGET='sandybridge,-xsaveopt;generic,clone_all;haswell,-rdrnd,base(1);x86-64-v4,-rdrnd,base(1)' julia-master example_targets.jl 
ERROR: LoadError: Base.PkgId(Base.UUID("7876af07-990d-54b4-ab0e-23690620f79a"), "Example") has not yet been precompiled for julia 1.12.0-DEV.531
Stacktrace:
 [1] error(::Base.PkgId, ::String, ::VersionNumber)
   @ Base ./error.jl:53
 [2] top-level scope
   @ /tmp/example_targets.jl:13
 [3] include
   @ ./Base.jl:559 [inlined]
 [4] exec_options(opts::Base.JLOptions)
   @ Base ./client.jl:325
 [5] _start()
   @ Base ./client.jl:533
in expression starting at /tmp/example_targets.jl:13
```
The `.julia/compiled/v1.12/Example` directory is empty.  Without `JULIA_CPU_TARGET`:
```console
% julia-master example_targets.jl                  
targets = Base.ImageTarget[haswell; flags=0; features_en=(sse3, pclmul, ssse3, fma, cx16, sse4.1, sse4.2, movbe, popcnt, aes, xsave, avx, f16c, fsgsbase, bmi, avx2, bmi2, sahf, lzcnt), haswell; flags=0; features_en=(sse3, pclmul, ssse3, fma, cx16, sse4.1, sse4.2, movbe, popcnt, aes, xsave, avx, f16c, fsgsbase, bmi, avx2, bmi2, sahf, lzcnt)]
```
Note that the target `haswell` is repeated twice.

**Comment by gbaraldi** [comment]
This isn't working :(

File: src/processor_arm.cpp [hunk 2]
```diff
@@ -1890,12 +1890,56 @@ const std::pair<std::string,std::string> &jl_get_llvm_disasm_target(void)
     return res;
 }
 
+#ifndef __clang_gcanalyzer__
 llvm::SmallVector<jl_target_spec_t, 0> jl_get_llvm_clone_targets(void)
 {
-    if (jit_targets.empty())
-        jl_error("JIT targets not initialized");
+
+    auto &cmdline = get_cmdline_targets();
```
Comment (giordano) [inline src/processor_arm.cpp:1897]: There seems to be error-prone duplication between this file and `src/processor_x86.cpp`, no?

Comment (gbaraldi) [reply src/processor_arm.cpp:1897]: Yes, these files mostly do the same thing, but not really and it's a mess.

File: src/processor_arm.cpp [hunk 3]
```diff
@@ -1890,12 +1890,56 @@ const std::pair<std::string,std::string> &jl_get_llvm_disasm_target(void)
...
+    check_cmdline(cmdline, true);
+    llvm::SmallVector<TargetData<feature_sz>, 0> image_targets;
+    for (auto &arg: cmdline) {
+        auto data = arg_target_data(arg, image_targets.empty());
+        image_targets.push_back(std::move(data));
+    }
+    auto ntargets = image_targets.size();
+    if (image_targets.empty())
+        jl_error("No targets specifiec");
```
Comment (IanButterworth) [inline src/processor_arm.cpp:1906]:
```suggestion
        jl_error("No targets specified");
```

File: src/processor_x86.cpp [hunk 4]
```diff
@@ -1102,13 +1104,82 @@ const std::pair<std::string,std::string> &jl_get_llvm_disasm_target(void)
...
+        static constexpr uint32_t clone_bf16[] = {Feature::avx512bf16};
+        for (auto fe: clone_bf16) {
+            if (!test_nbit(features0, fe) && test_nbit(t.en.features, fe)) {
+                t.en.flags |= JL_TARGET_CLONE_BFLOAT16;
+                break;
+            }
+        }
+    }
+    if (image_targets.empty())
+        jl_error("No targets specifiec");
```
Comment (IanButterworth) [inline src/processor_x86.cpp:1180]:
```suggestion
        jl_error("No targets specified");
```

**Comment by IanButterworth** [comment]
Recording here so this isn't lost on slack.

There's a ~27% increase in pkgimage size with this PR.

master (via juliaup)
```
julia> using PkgCacheInspector 

julia> info_cachefile("Pkg")
Contents of /Users/ian/.julia/juliaup/julia-nightly/share/julia/compiled/v1.12/Pkg/tUTdb_myMUk.dylib:
  modules: Any[Pkg.MiniProgressBars, Pkg.GitTools, Pkg.PlatformEngines, Pkg.Versions, Pkg.Registry, Pkg.Resolve, Pkg.Types.FuzzySorting, Pkg.Types, Pkg.BinaryPlatforms, Pkg.Artifacts, Pkg.Operations, Pkg.API, Pkg.REPLMode, Pkg]
  init order: Any[Pkg]
  529 external methods
  16850 new specializations of external methods (Base 79.9%, Base.Broadcast 7.6%, Base.Sort 3.7%, ...)
  1585 external methods with new roots
  33049 external targets
  24615 edges
  file size:   47139456 (44.956 MiB)
  Segment sizes (bytes):
    system:      19914892 ( 47.10%)
    isbits:      19778248 ( 46.78%)
    symbols:        80497 (  0.19%)
    tags:          336949 (  0.80%)
    relocations:  2082887 (  4.93%)
    gvars:          41200 (  0.10%)
    fptrs:          44632 (  0.11%)
  Image targets: 
    generic; flags=0; features_en=()
```

This PR built locally on MacOS with the same buildkite targets
```
JULIA_CPU_TARGET=generic;cortex-a57;thunderx2t99;carmel,clone_all;apple-m1,base(3);neoverse-512tvb,base(3)
```
```
julia> info_cachefile("Pkg")
Contents of /Users/ian/Documents/GitHub/julia/usr/share/julia/compiled/v1.12/Pkg/tUTdb_p5Nph.dylib:
  modules: Any[Pkg.MiniProgressBars, Pkg.GitTools, Pkg.PlatformEngines, Pkg.Versions, Pkg.Registry, Pkg.Resolve, Pkg.Types.FuzzySorting, Pkg.Types, Pkg.BinaryPlatforms, Pkg.Artifacts, Pkg.Operations, Pkg.API, Pkg.REPLMode, Pkg]
  init order: Any[Pkg]
  529 external methods
  16819 new specializations of external methods (Base 79.9%, Base.Broadcast 7.6%, Base.Sort 3.7%, ...)
  1581 external methods with new roots
  32968 external targets
  24583 edges
  file size:   59591872 (56.831 MiB)
  Segment sizes (bytes):
    system:      19918588 ( 47.19%)
    isbits:      19712700 ( 46.70%)
    symbols:        76445 (  0.18%)
    tags:          336483 (  0.80%)
    relocations:  2080035 (  4.93%)
    gvars:          41216 (  0.10%)
    fptrs:          44520 (  0.11%)
  Image targets: 
    generic; flags=0; features_en=()
    cortex-a57; flags=0; features_en=(crc)
    thunderx2t99; flags=0; features_en=(aes, sha2, crc, lse, rdm, v8_1a)
    carmel; flags=0; features_en=(aes, sha2, crc, lse, fullfp16, rdm, ccpp, v8_1a, v8_2a)
    apple-m1; flags=0; features_en=(aes, sha2, crc, lse, fullfp16, rdm, jsconv, complxnum, rcpc, ccpp, sha3, dotprod, fp16fml, dit, rcpc-immo, flagm, sb, ccdp, altnzcv, fptoint, v8_1a, v8_2a, v8_3a, v8_4a, v8_5a)
    neoverse-512tvb; flags=32; features_en=()
```

**Comment by StefanKarpinski** [comment]
That's not bad for being having efficient code for multiple platforms! This is definitely something we'll want for serving pkgimgs.

**Comment by gbaraldi** [comment]
There is a windows issue somehow :\

**Comment by IanButterworth** [comment]
If there's no clear reason for the windows issue is it worth disabling on windows and merging this to get it tested?

**Comment by gbaraldi** [comment]
I'm not super confortable merging it because that windows issue doesn't seem to be simple.

**Comment by vchuravy** [comment]
```
Error During Test at C:\buildkite-agent\builds\win2k22-amdci6-6\julialang\julia-master\julia-b1c3d856db\share\julia\test\precompile_utils.jl:4
  Got exception outside of a @test
  f35014 not found in sysimg
  Stacktrace:
```

That looks like a symbol visibility issue? @pchintalapudi might have an idea.

**Comment by vchuravy** [comment]
E.g. maybe something similar to https://github.com/JuliaLang/julia/pull/51530 for just the ccallable?

**Comment by giordano** [comment]
What's the status of this PR?  I had a look a few days ago, my understanding is that a local native build of Julia which doesn't set `JULIA_CPU_TARGET` during the compilation of the sysimage can't generate pkgimages for different (or smaller than "native" for the host?) ISAs, while setting `JULIA_CPU_TARGET` during sysimage building allows compiling pkgimages for targets compatible with the those included in the sysimage.  Is this an accurate description?  If so, I feel like this should be better documented, because it isn't very obvious.

**Comment by gbaraldi** [comment]
Yes, basically pkgimages can only be more specific than the base system image so if the the base sysimage is native, then the pkgimages are also native.

**Review by vchuravy — APPROVED** [review]

File: test/precompile.jl [hunk 5]
```diff
@@ -1963,6 +1963,19 @@ precompile_test_harness("Test flags") do load_path
...
+        pkg = Base.identify_package("Test")  # but we need better target reflection to make a better one.
+        cachefiles = Base.find_all_in_cache_path(pkg)
+        pkgpath = Base.locate_package(pkg)
+        idx = findfirst(cachefiles) do cf
+            Base.stale_cachefile(pkgpath, cf) !== true
+        end
+        targets = Base.parse_image_targets(Base.parse_cache_header(cachefiles[idx])[7])
+        @test length(targets) > 1
+    end
+end
```
Comment (IanButterworth) [inline test/precompile.jl:1977]:
```suggestion
end
```

**Comment by Octogonapus** [comment]
Great to have this improved but I agree with Mose that this is not clear:

> basically pkgimages can only be more specific than the base system image

Is an error returned if the user tries to make a package image for a totally different target than the base image? I don't fully understand this code but it seems that there is no check for that case. Would it be feasible to implement such a check? Otherwise this feels like a footgun. I could also update the [docs here](https://docs.julialang.org/en/v1/manual/environment-variables/#JULIA_CPU_TARGET) and [here](https://docs.julialang.org/en/v1/devdocs/pkgimg/#pkgimgs-multi-versioning) but I don't think that would be enough.

**Comment by KristofferC** [comment]
Sorry, I should have written something. It was discussed that this was too big of a thing to backport as late as it was in the release cycle. Now that it has been on master for a while, perhaps that can be reconsidered.

**Comment by giordano** [comment]
Since backporting this PR would fix #56177, I'm inclined to do it.

**Comment by gbaraldi** [comment]
Talking with @kpamnany. It would be good to have this in 1.10. It has been extensively tested in both master and 1.11 so I don't see a big issue

**Comment by IanButterworth** [comment]
Might this slow down 1.10 load/precompile speed?

**Comment by vchuravy** [comment]
Precompile yes, since there is more work being done, but only when the user is actually requesting it. Loading should not be impacted.

**Comment by DilumAluthge** [comment]
This doesn't backport cleanly to 1.10.

@gbaraldi Would you be able to make a manual backport to 1.10?

**Comment by giordano** [comment]
It doesn't apply cleanly because [it's ***already*** there](https://github.com/JuliaLang/julia/commit/7a28ad5aeb681e7631b360e2998aa998192d1601), no?

**Comment by IanButterworth** [comment]
I think Backporter is confused because the title truncates on that commit. https://github.com/KristofferC/Backporter/pull/20

**Pull request merged**